source/module_base/ndarray.h

```diff
@@ -0,0 +1,284 @@
+            shape_ = std::move(other.shape_);
+        }
+        return *this;
+    }
+
+    /**
+     * @brief == operator
+     * 
+     * @param other 
+     * @return true if the data and shape are the same
+     * @return false otherwise
+     */
+    bool operator==(const NDArray& other) const
+    {
+        return data_ == other.data_ && shape_ == other.shape_;
+    }
+    /**
+     * @brief != operator
+     * 
+     * @param other 
+     * @return true if the data and shape are different
+     * @return false otherwise
+     */
+    bool operator!=(const NDArray& other) const
+    {
+        return !(*this == other);
+    }
+    // other operators are not generally supported
+
+    // element access
+    /**
+     * @brief at function
+     * 
+     * @tparam Args 
+     * @param args indices of the element
+     * @return T& 
+     */
+    template<typename... Args>
+    T& at(Args... args)
+    {
+        size_t idx = index(args...);
+        return data_.at(idx);
+    }
+    template<typename... Args>
+    const T& at(Args... args) const
+    {
+        size_t idx = index(args...);
+        return data_.at(idx);
+    }
+    /**
+     * @brief [] operator
+     * 
+     * @tparam Args 
+     * @param args indices of the element
+     * @return T& 
+     */
+    template<typename... Args>
+    T& operator()(Args... args)
+    {
+        size_t idx = index(args...);
+        return data_.at(idx);
+    }
+    /**
+     * @brief const [] operator
+     * 
+     * @tparam Args 
+     * @param args indices of the element
+     * @return const T& 
+     */
+    template<typename... Args>
+    const T& operator()(Args... args) const
+    {
+        size_t idx = index(args...);
+        return data_.at(idx);
+    }
+    // front
+    T& front() { return data_.front(); }
+    const T& front() const { return data_.front(); }
+    // back
+    T& back() { return data_.back(); }
+    const T& back() const { return data_.back(); }
+    // data
+    T* data() { return data_.data(); }
+    const T* data() const { return data_.data(); }
+
+    // iterators
+    // iterators on the whole data
+    T* begin() { return data_.data(); }
+    T* end() { return data_.data() + data_.size(); }
+    const T* cbegin() const { return data_.data(); }
+    const T* cend() const { return data_.data() + data_.size(); }
+    // iterators on different dimensions
+    
+    // capacity
+    // size
+    size_t size() const { return data_.size(); }
+    size_t size(const size_t& dim) const
+    {
+        assert(dim < shape_.size());
+        return shape_[dim];
+    }
+    /**
+     * @brief change the capacity of the data
+     * 
+     * @param new_cap new capacity
+     */
+    void reserve(size_t new_cap) { data_.reserve(new_cap); }
+    /**
+     * @brief change the capacity of the data and fill the new elements with val
+     * 
+     * @param new_cap new capacity
+     * @param val default value
+     */
+    void reserve(size_t new_cap, const T& val)
```

Comment:
this function looks dubious... looks like it may break the consistency between data & shape! Moreover, I think there's no need to reserve before resize for a std::vector (if the target size is the same).

Reply:
yes I also doubt the necessity of implemententation of this function, seems the memory space the instance occupies can be determined at call of constructor, and no explicit need to call this function...